Widgets/VPNWidget.swift

```diff
@@ -0,0 +1,291 @@
+    let location: String
+
+    internal init(date: Date, status: VPNStatus = .notConfigured, location: String) {
+        self.date = date
+        self.status = status
+        self.location = location
+    }
+}
+
+class VPNStatusTimelineProvider: TimelineProvider {
+
+    typealias Entry = VPNStatusTimelineEntry
+
+    func placeholder(in context: Context) -> VPNStatusTimelineEntry {
+        return VPNStatusTimelineEntry(date: Date(), status: .status(.connected), location: "Los Angeles, CA")
+    }
+
+    func getSnapshot(in context: Context, completion: @escaping (VPNStatusTimelineEntry) -> Void) {
+        let entry = VPNStatusTimelineEntry(date: Date(), status: .status(.connected), location: "Los Angeles, CA")
+        completion(entry)
+    }
+
+    func getTimeline(in context: Context, completion: @escaping (Timeline<VPNStatusTimelineEntry>) -> Void) {
+        NETunnelProviderManager.loadAllFromPreferences { managers, error in
+            let defaults = UserDefaults.networkProtectionGroupDefaults
+            let location = defaults.string(forKey: NetworkProtectionUserDefaultKeys.lastSelectedServer) ?? "Unknown Location"
+            let expiration = Date().addingTimeInterval(TimeInterval.minutes(5))
+
+            if error != nil {
+                let entry = VPNStatusTimelineEntry(date: expiration, status: .error, location: location)
+                let timeline = Timeline(entries: [entry], policy: .atEnd)
+                completion(timeline)
+                return
+            }
+
+            guard let manager = managers?.first else {
+                let entry = VPNStatusTimelineEntry(date: expiration, status: .notConfigured, location: location)
+                let timeline = Timeline(entries: [entry], policy: .atEnd)
+                completion(timeline)
+                return
+            }
+
+            let status = manager.connection.status
+            let entry = VPNStatusTimelineEntry(date: expiration, status: .status(status), location: location)
+            let timeline = Timeline(entries: [entry], policy: .atEnd)
+
+            completion(timeline)
+        }
+    }
+}
+
+extension NEVPNStatus {
+    var description: String {
+        switch self {
+        case .connected: return "Connected"
+        case .connecting: return "Connecting"
+        case .disconnected: return "Disconnected"
+        case .disconnecting: return "Disconnecting"
+        case .invalid: return "Invalid"
+        case .reasserting: return "Reasserting"
+        default: return "Unknown Status"
+        }
+    }
+
+    var isConnected: Bool {
+        switch self {
+        case .connected, .connecting, .reasserting: return true
+        case .disconnecting, .disconnected: return false
+        default: return false
+        }
+    }
+}
+
+@available(iOSApplicationExtension 17.0, *)
+struct VPNStatusView: View {
+    @Environment(\.widgetFamily) var family: WidgetFamily
+    var entry: VPNStatusTimelineProvider.Entry
```

Comment:
I see that a lot of this has a very synchronous data flow. I’m not sure where the limitations lie with that but let’s have a quick chat about it now I’ve seen the PR.